contracts/smart-account/modules/MultichainValidator.sol

```diff
@@ -47,27 +39,36 @@ contract MultichainECDSAValidator is EcdsaOwnershipRegistryModule {
 
         //otherwise it is a multichain signature
         (
+            uint48 validUntil,
+            uint48 validAfter,
             bytes32 merkleTreeRoot,
             bytes32[] memory merkleProof,
             bytes memory multichainSignature
-        ) = abi.decode(moduleSignature, (bytes32, bytes32[], bytes));
+        ) = abi.decode(
+                moduleSignature,
+                (uint48, uint48, bytes32, bytes32[], bytes)
+            );
 
-        if (!MerkleProof.verify(merkleProof, merkleTreeRoot, userOpHash)) {
+        //make a leaf out of userOpHash, validUntil and validAfter
+        bytes32 leaf = keccak256(
+            abi.encodePacked(validUntil, validAfter, userOpHash)
+        );
+
+        if (!MerkleProof.verify(merkleProof, merkleTreeRoot, leaf)) {
             revert("Invalid UserOp");
         }
 
-        // reconstruct hash = all the userOp fileds except nonce + merkleTreeRoot that is based on chainId, nonce, address(this)
-        /* bytes32 multichainHash = keccak256(
-            abi.encode(getChainAgnosticUserOpHash(userOp), merkleTreeRoot)
-        ); */
-
         return
             _verifySignature(
                 merkleTreeRoot,
                 multichainSignature,
                 address(uint160(sender))
             )
-                ? VALIDATION_SUCCESS
+                ? _packValidationData(
+                    false, //sigVerificationFailed = false
+                    validUntil == 0 ? type(uint48).max : validUntil,
+                    validAfter
+                )
                 : SIG_VALIDATION_FAILED;
```

Comment:
why we are not returning packed validation data in case of failed signature validation? Is it not used in EntryPoint ?

Reply:
in case signature validation failed, there's no need to spend gas and pack validUntil and validAfter. they do not matter if the sig itself is wrong.
thus, having SIG_VALIDATION_FAILED value just 1 and returning it is enough. 
https://github.com/eth-infinitism/account-abstraction/blob/12be13e2e97b763e1ef294602b3f2072bc301443/contracts/samples/SimpleAccount.sol#L104